# 01 - Prima esplorazione del dataset

This is a dataset of Spotify tracks over a range of 114 different genres. 
Each track has some audio features associated with it. The data is in CSV 
format which is tabular and can be loaded quickly. https://www.kaggle.com/datasets/maharshipandya/-spotify-tracks-dataset?select=dataset.csv. 
Exploring the dataset to acquire information

## Imports

In [1]:
import pandas as pd

## Raw csv load from data/raw

In [2]:
spotify = pd.read_csv("../data/raw/spotify.csv")

## Rows and columns

In [3]:
spotify.shape

(114000, 21)

In [4]:
spotify.head()

,Unnamed: 0,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,...,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.676,0.4610,...,-6.746,0,0.1430,0.0322,0.000001,0.3580,0.715,87.917,4,acoustic
1,1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.420,0.1660,...,-17.235,1,0.0763,0.9240,0.000006,0.1010,0.267,77.489,4,acoustic
2,2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.438,0.3590,...,-9.734,1,0.0557,0.2100,0.000000,0.1170,0.120,76.332,4,acoustic
3,3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,Can't Help Falling In Love,71,201933,False,0.266,0.0596,...,-18.515,1,0.0363,0.9050,0.000071,0.1320,0.143,181.740,3,acoustic
4,4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.618,0.4430,...,-9.681,1,0.0526,0.4690,0.000000,0.0829,0.167,119.949,4,acoustic


In [5]:
spotify.head().T

,0,1,2,3,4
Unnamed: 0,0,1,2,3,4
track_id,5SuOikwiRyPMVoIQDJUgSV,4qPNDBW1i3p13qLCt0Ki3A,1iJBSr7s7jYXzM8EGcbK5b,6lfxq3CG4xtTiEg7opyCyx,5vjLSffimiIP26QG5WcN2K
artists,Gen Hoshino,Ben Woodward,Ingrid Michaelson;ZAYN,Kina Grannis,Chord Overstreet
album_name,Comedy,Ghost (Acoustic),To Begin Again,Crazy Rich Asians (Original Motion Picture Sou...,Hold On
track_name,Comedy,Ghost - Acoustic,To Begin Again,Can't Help Falling In Love,Hold On
popularity,73,55,57,71,82
duration_ms,230666,149610,210826,201933,198853
explicit,False,False,False,False,False
danceability,0.676,0.42,0.438,0.266,0.618
energy,0.461,0.166,0.359,0.0596,0.443


Dai primi dati è possibile osservare che le colonne hanno tutte un nome (tranne la prima Unnamed:0), 
e ora i dati sembrano sensati

## Column types and missing values

In [6]:
spotify.info()

<class 'pandas.DataFrame'>
RangeIndex: 114000 entries, 0 to 113999
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        114000 non-null  int64  
 1   track_id          114000 non-null  str    
 2   artists           113999 non-null  str    
 3   album_name        113999 non-null  str    
 4   track_name        113999 non-null  str    
 5   popularity        114000 non-null  int64  
 6   duration_ms       114000 non-null  int64  
 7   explicit          114000 non-null  bool   
 8   danceability      114000 non-null  float64
 9   energy            114000 non-null  float64
 10  key               114000 non-null  int64  
 11  loudness          114000 non-null  float64
 12  mode              114000 non-null  int64  
 13  speechiness       114000 non-null  float64
 14  acousticness      114000 non-null  float64
 15  instrumentalness  114000 non-null  float64
 16  liveness          114000 non-nu

In [7]:
spotify.isna().sum()

Unnamed: 0          0
track_id            0
artists             1
album_name          1
track_name          1
popularity          0
duration_ms         0
explicit            0
danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
time_signature      0
track_genre         0
dtype: int64

Da questi dati si osserva che ci sono 1 missing values per artists, 
1 per album_name e 1 per track_name. Molto pochi rispetto al totale. 
I tipi sembrano sensati e coerenti.

## Numeric columns stats

In [8]:
spotify[["popularity", "duration_ms", "tempo", "time_signature"]].describe()

,popularity,duration_ms,tempo,time_signature
count,114000.000000,1.140000e+05,114000.000000,114000.000000
mean,33.238535,2.280292e+05,122.147837,3.904035
std,22.305078,1.072977e+05,29.978197,0.432621
min,0.000000,0.000000e+00,0.000000,0.000000
25%,17.000000,1.740660e+05,99.218750,4.000000
50%,35.000000,2.129060e+05,122.017000,4.000000
75%,50.000000,2.615060e+05,140.071000,4.000000
max,100.000000,5.237295e+06,243.372000,5.000000


I min e max della popularity sembrano sensati, e la scala va da 0 a 100.
La media è di 2.2 min per la durata, mentre la cosa che sembra strana 
è il minimo di durata (0), significa che esistono una o più canzoni da 0 minuti.
Anche in questo caso il tempo min è 0, cosa molto strana.
time signature va da 0 a 5, strano visto che il valore 0 è impossibili musicalmente.


## Categories descriptions

In [9]:
spotify.select_dtypes(include="object").nunique()

/var/folders/bc/2zhpmgw10sbg30n8kjvgkhph0000gn/T/ipykernel_60312/2875575617.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  spotify.select_dtypes(include="object").nunique()


track_id       89741
artists        31437
album_name     46589
track_name     73608
track_genre      114
dtype: int64

In [10]:
spotify["track_genre"].value_counts().head(20)

track_genre
acoustic         1000
afrobeat         1000
alt-rock         1000
alternative      1000
ambient          1000
anime            1000
black-metal      1000
bluegrass        1000
blues            1000
brazil           1000
breakbeat        1000
british          1000
cantopop         1000
chicago-house    1000
children         1000
chill            1000
classical        1000
club             1000
comedy           1000
country          1000
Name: count, dtype: int64

In [11]:
spotify["artists"].value_counts().head(10)

artists
The Beatles        279
George Jones       271
Stevie Wonder      236
Linkin Park        224
Ella Fitzgerald    222
Prateek Kuhad      217
Feid               202
Chuck Berry        190
Håkan Hellström    183
OneRepublic        181
Name: count, dtype: int64

Il numero più alto di categorie uniche è l'id track ovviamente.

Tra i primi 20 generi, ci sono 1000 tracce per genere. Molto strano,probabilmente in questo database sono state prese 1000 tracce per genere (infatti ci sono 114 generi, moltiplicati per mille corrispondono esattamente al numero di tracce).

Tra i top 10 artisti non risultano collaborazioni, ma non posso escluderle più in basso nella distribuzione.

## Duplicates

In [12]:
spotify.duplicated().sum()

np.int64(0)

In [13]:
spotify.duplicated(subset=["track_id"]).sum()

np.int64(24259)

In [14]:
spotify.duplicated(subset=["track_name", "artists"]).sum()

np.int64(32656)

Apparentemente, ci sono 24259 tracce con lo stesso id, ciò significa che probabilmente
più tracce con lo stesso id sono state assegnate a generi diversi o eventualmente
cantate da artisti diversi.

La stessa cosa può valere per i duplicati con lo stesso artista e stesso nome della traccia

**SINTESI FINALE**
- Il dataset contiene 114k righe e 21 colonne;
- Struttura: (114.000 x 21). 4 colonne testuali (track_id, artists, album_name, track_name, track_genre), 5 colonne numeriche (popularity, duration_ms, feature audio come danceability/energy/valence..., ecc), 1 colonna booleana (explicit), 1 colonna anonima "Unnamed:0" da droppare;
- Tre missing (non critici): un artist missing, una track_name missing e un album_name missing. Considerata la quantità di righe, avere un missing in queste categorie non dovrebbe interferire con l'analisi;

**todo**
- Droppare la colonna anonima "Unnamed: 0"
- Decidere cosa fare dei 3 missing (drop? fill?);
- Investigare canzoni con durata = 0;
- Investigare canzoni con tempo = 0;
- Investigare righe con time_signature = 0;
- Gestire i duplicati di track_id (mantenere tutti, oppure deduplicare per analisi specifiche)